TimeOfFlight/src/TOFEstimators.cc

```diff
@@ -33,442 +43,591 @@
 #include <gsl/gsl_rng.h>
 #include <gsl/gsl_randist.h>
 
-//---- ROOT -----
-#include "TH2F.h"
-
 using namespace lcio ;
 using namespace marlin ;
 using namespace TOFUtils ;
 
 TOFEstimators aTOFEstimators ;
 
 
-
 TOFEstimators::TOFEstimators() : Processor("TOFEstimators") {
+    _description = "TOFEstimators compute some estimators for the time of flight from calorimeter hits" ;
+
+    // register steering parameters: name, description, class-variable, default value
+    registerInputCollection( LCIO::RECONSTRUCTEDPARTICLE,
+                            "ReconstructedParticleCollection",
+                            "Name of the ReconstructedParticle collection",
+                            _colNamePFO,
+                            std::string("PandoraPFOs") );
+
+    registerProcessorParameter( "MaxLayerNumber",
+                            "Use only calorimeter hits up to MaxLayerNumber in TOF estimators",
+                            _maxLayerNum,
+                            int(100) );
+
+    registerProcessorParameter( "CylRadius",
+                            "Cut-off hits further away from the shower core than this raduius for TOF calculation",
+                            _cylRadiusCut,
+                            double(5.) );
+
+
+    registerProcessorParameter( "TimeResolution",
+                            "Assumed time resolution per hit in ps",
+                            _resolution,
+                            float(0.) );
+
+    registerProcessorParameter( "ProcessorVersion",
+                            "Legacy or new TOFEstimator",
+                            _procVersion,
+                            std::string("idr") );
+}
 
-  // modify processor description
-  _description = "TOFEstimators compute some estimators for the time of flight from calorimeter hits" ;
 
-  // register steering parameters: name, description, class-variable, default value
-  registerInputCollection( LCIO::RECONSTRUCTEDPARTICLE,
-			   "ReconstructedParticleCollection" , 
-			   "Name of the ReconstructedParticle collection"  ,
-			   _colNamePFO ,
-			   std::string("PandoraPFOs")
-    );
+void TOFEstimators::init() {
+    Global::EVENTSEEDER->registerProcessor(this);
+    if ( _procVersion == "idr" ){
+        // initialize gsl random generator
+        _rng = gsl_rng_alloc(gsl_rng_ranlxs2);
+        _TOFNames = {"TOFFirstHit", "TOFClosestHits", "TOFClosestHitsError", "TOFFlightLength",  "TOFLastTrkHit" , "TOFLastTrkHitFlightLength"};
+    }
+    else if (_procVersion == "dev"){
+        _smearing.param( std::normal_distribution<double>::param_type(0., _resolution/1000.) );
+        _TOFNames = {"TOFClosest", "TOFFastest", "TOFCylFit", "TOFClosestFit", "FlightLength", "MomAtCalo"};
 
+        const auto& detector = dd4hep::Detector::getInstance();
+        detector.field().magneticField({0., 0., 0.}, _bField);
+    }
+    else { throw std::string("Invalid ProcessorVersion parameter passed!!!\n Viable options are: idr (default), dev"); }
```

Comment:
```suggestion
    else { 
        throw EVENT::Exception(std::string("Invalid ProcessorVersion parameter passed: \'") + _procVersion + std::string("\'\n Viable options are idr (default), dev"); 
    }
```
Sorry, missed this before. The important part here is to throw an exception that Marlin/LCIO can properly handle, and also give the user some context when it crashes. Simply throwing a string here potentially gets past all that and simply gives a crash with a stack-trace which is somewhat harder to decipher.
If you include the passed parameter in the exception message is a minor detail in the end.